File: 0091-decode-ways/0091-decode-ways.cpp
// class Solution {
// public:
//     int rec(int idx, string s, int n, vector<int> &dp) {
//         if(dp[idx] != -1) return dp[idx];
//         if(s[idx] == '0') return dp[idx] = 0;
//         if(idx == n) return dp[idx] = 1;
//         int res = rec(idx + 1, s, n, dp);
//         if(idx < n - 1 && (s[idx] == '1' || (s[idx] == '2' && s[idx + 1] < '7'))) res += rec(idx + 2, s, n, dp);
//         return dp[idx] = res;
//     }

//     int numDecodings(string s) {
//         int n = s.size();
//         vector<int> dp(n + 1, -1);
//         return s.empty() ? 0 : rec(0, s, n, dp);
//     }
// };

class Solution {
public:
    int numDecodings(string s) {
        int n = s.size();
        vector<int> dp(n + 1, -1);
        dp[n] = 1;

        for(int i = n - 1; i >= 0; i--) {
            if(s[i] == '0') dp[i] = 0;
            else {
                dp[i] = dp[i + 1];
                if(i < n - 1 && (s[i] == '1' || (s[i] == '2' && s[i + 1] < '7'))) dp[i] += dp[i + 2];
            }
        }

        return s.empty() ? 0 : dp[0];
    }
};

File: 0091-decode-ways/0091-decode-ways.java
class Solution {
    private int memo(int idx, int n, String s, int[] dp) {
        if(idx == n) return 1;  // Base case: reached end successfully
        if(s.charAt(idx) == '0') return 0; // Invalid: can't start with 0
        if(dp[idx] != -1) return dp[idx];
        
        // Single digit decoding
        int res = memo(idx + 1, n, s, dp);
        
        // Two digit decoding (10-26)
        if(idx < n - 1 && (s.charAt(idx) == '1' || 
           (s.charAt(idx) == '2' && s.charAt(idx + 1) < '7'))) {
            res += memo(idx + 2, n, s, dp);
        }
        
        return dp[idx] = res;
    }
    
    public int numDecodings(String s) {
        int n = s.length();
        int[] dp = new int[n];
        Arrays.fill(dp, -1);
        
        return memo(0, n, s, dp);
    }
}